pkg/operator/options/options.go

```diff
@@ -175,11 +176,25 @@ func FromContext(ctx context.Context) *Options {
 	return retval.(*Options)
 }
 
+var clusterIDRegex = regexp.MustCompile(`^([^.-]+(?:-[^.-]+)*?)(?:-[^.-]+)?\.`)
+
 // getAKSClusterID returns cluster ID based on the DNS prefix of the cluster.
-// The logic comes from AgentBaker and other places, originally from aks-engine
-// with the additional assumption of DNS prefix being the first 33 chars of FQDN
 func getAKSClusterID(apiServerFQDN string) string {
-	dnsPrefix := apiServerFQDN[:33]
+	// TODO: This doesn't work for clusters with BYO private dns zone with fqdnSubdomain. For those clusters
+	//       we need to hash the emptyString to get the cluster ID (that's what AKS RP does), but
+	//       you can't actually tell by examining a fqdn record if it's using dnsPrefix or fqdnSubdomain.
+	//       You ALMOST could, by looking for <dnsprefix>-<8 character random string>.privatelink.<region>.azmk8s.io, but
+	//       There's nothing stopping users from making their dns subdomain with an 8 character random string in the same
+	//       location that AKS put it.
+	//       I think it would be cleaner to just read this value directly from the nsg in the MC RG, rather than
```

Comment:
If there's agreement on this I can just delete this function in a followup PR and we can source it from the NSG name directly instead. The two main points of awkwardness there are:
1. It means that we have to add it to options later as opposed to computing it inline as part of options parse, which is a bit uglier.
2. If there are any cases where a cluster doesn't have an NSG, it wouldn't work. I don't think that ever happens though.

Reply:
We need the NSG provider in the case that the cluster is BYO VNET, because we have to attach the NSG to the NIC ourselves in that case.

Of course, if we had a durable way to generate the NSG name (using a reliable cluster ID) we could have done that instead, but given that we already have access to the node resource group, and we don't (as far as I can tell) have a reliable way to get cluster ID due to the points I called out in this code comment, we do need the NSG provider so my thinking is more that we might as well use it all the time as opposed to only in BYO VNET.

I'll follow up with a PR to just remove this method entirely... may hold off on merging this one in the meantime.